### PR TITLE
Support case-insensitive filesystems when pruning files after update

### LINKED ISSUE #195: Core_Command::cleanup_extra_files() incorrectly deletes files on case-insensitive filesystems when an update changes the case of a file
## Bug Report

- [X] Yes, I reviewed the [contribution guidelines](https://make.wordpress.org/cli/handbook/contributing/).
- [x] Yes, more specifically, I reviewed the guidelines on [how to write clear bug reports](https://make.wordpress.org/cli/handbook/bug-reports/).

**Describe the current, buggy behavior**

During the WP 5.9-beta1 release party, it was discovered that a number of files in the `Requests` library has their filenames changed only in case (e.g., `wp-includes/Requests/IRI.php` -> `wp-includes/Requests/Iri.php`.

When updating to 5.9-beta 1 via wp-cli on a machine with a case in-sensitive filesystem (e.g., Windows), those files are deleted in `Core_Command::cleanup_extra_files()` and that results in PHP fatal errors whenever core tries to use the Requests library.

**Describe what you would expect as the correct outcome**

I expect the detection of files to be deleted to be done in a case-insensitive manner.

**Let us know what environment you are running this on**

Windows 10

```
Cleaning up files...
File removed: wp-includes/Requests/Transport/fsockopen.php
File removed: wp-includes/Requests/Transport/cURL.php
File removed: wp-includes/Requests/Hooker.php
File removed: wp-includes/Requests/IPv6.php
File removed: wp-includes/Requests/Exception/Transport/cURL.php
File removed: wp-includes/Requests/Exception/HTTP.php
File removed: wp-includes/Requests/Exception/HTTP/502.php
File removed: wp-includes/Requests/Exception/HTTP/Unknown.php
File removed: wp-includes/Requests/Exception/HTTP/412.php
File removed: wp-includes/Requests/Exception/HTTP/408.php
File removed: wp-includes/Requests/Exception/HTTP/431.php
File removed: wp-includes/Requests/Exception/HTTP/501.php
File removed: wp-includes/Requests/Exception/HTTP/500.php
File removed: wp-includes/Requests/Exception/HTTP/407.php
File removed: wp-includes/Requests/Exception/HTTP/416.php
File removed: wp-includes/Requests/Exception/HTTP/428.php
File removed: wp-includes/Requests/Exception/HTTP/406.php
File removed: wp-includes/Requests/Exception/HTTP/504.php
File removed: wp-includes/Requests/Exception/HTTP/411.php
File removed: wp-includes/Requests/Exception/HTTP/414.php
File removed: wp-includes/Requests/Exception/HTTP/511.php
File removed: wp-includes/Requests/Exception/HTTP/410.php
File removed: wp-includes/Requests/Exception/HTTP/403.php
File removed: wp-includes/Requests/Exception/HTTP/400.php
File removed: wp-includes/Requests/Exception/HTTP/505.php
File removed: wp-includes/Requests/Exception/HTTP/413.php
File removed: wp-includes/Requests/Exception/HTTP/404.php
File removed: wp-includes/Requests/Exception/HTTP/306.php
File removed: wp-includes/Requests/Exception/HTTP/304.php
File removed: wp-includes/Requests/Exception/HTTP/405.php
File removed: wp-includes/Requests/Exception/HTTP/429.php
File removed: wp-includes/Requests/Exception/HTTP/417.php
File removed: wp-includes/Requests/Exception/HTTP/409.php
File removed: wp-includes/Requests/Exception/HTTP/402.php
File removed: wp-includes/Requests/Exception/HTTP/418.php
File removed: wp-includes/Requests/Exception/HTTP/305.php
File removed: wp-includes/Requests/Exception/HTTP/415.php
File removed: wp-includes/Requests/Exception/HTTP/401.php
File removed: wp-includes/Requests/Exception/HTTP/503.php
File removed: wp-includes/Requests/IRI.php
File removed: wp-includes/Requests/IDNAEncoder.php
File removed: wp-includes/Requests/SSL.php
File removed: wp-includes/Requests/Proxy/HTTP.php
File removed: wp-includes/blocks/post-content/editor.min.css
File removed: wp-includes/blocks/post-content/editor.css
File removed: wp-includes/blocks/post-content/editor-rtl.css
File removed: wp-includes/blocks/post-content/editor-rtl.min.css
File removed: wp-includes/blocks/heading/editor.min.css
File removed: wp-includes/blocks/heading/editor.css
File removed: wp-includes/blocks/heading/editor-rtl.css
File removed: wp-includes/blocks/heading/editor-rtl.min.css
File removed: wp-includes/blocks/query-title/editor.min.css
File removed: wp-includes/blocks/query-title/editor.css
File removed: wp-includes/blocks/query-title/editor-rtl.css
File removed: wp-includes/blocks/query-title/editor-rtl.min.css
File removed: wp-includes/blocks/tag-cloud/editor.min.css
File removed: wp-includes/blocks/tag-cloud/editor.css
File removed: wp-includes/blocks/tag-cloud/editor-rtl.css
File removed: wp-includes/blocks/tag-cloud/editor-rtl.min.css
59 files cleaned up.
```

**Provide a possible solution**

I think it should be relatively easy to fix by doing case folding (e.g., convert everything to lower case) before doing:

`$files_to_remove = array_diff( array_keys( $old_checksums ), array_keys( $new_checksums ) );`

### PR DESCRIPTION
**Pull request by pbiron**
Fixes #195

### CODE REVIEW

**Review by schlessera — APPROVED** [review]

**Comment by schlessera** [comment]
It's getting ever more complicated. Right now, the current code fails on some PHP versions and succeeds on others:
Allow for warnings to be produced when verifying checksums after a downgrad  2021-12-06 at 11 53 16 AM

**Comment by pbiron** [comment]
Also, given the discussion in [Slack #core](https://wordpress.slack.com/archives/C02RQBWTW/p1638803229013100) about reverting the Requests 2.0.0 changes in WP 5.9 (and dealing with the complications they are causing, in 6.0) this PR probably doesn't need to merged right now...

**Comment by pbiron** [comment]
And here's more weirdness.

Over the weekend, @wojsmol incorporated this PR into one on his fork so that we could see the behat tests run before you had enabled them to run here.  And the test was failing, see https://github.com/wojsmol/core-command/pull/2...in a way that suggests the PR isn't correct.  See, for example, https://github.com/wp-cli/core-command/runs/4433324549?check_suite_focus=true#step:13:40.

But the failures here seem total different (and maybe even unrelated to the PR itself).

**Comment by wojsmol** [comment]
PR on my fork is against branch of https://github.com/wp-cli/core-command/pull/197 not against core-command master branch.

**Comment by schlessera** [comment]
@pbiron I'm currently looking into getting this to work. I have trouble understanding how the PR is supposed to work, though.

Right now, what it does is that it avoids deleting files from `$old` when a file on `$new` has the same name or a similar name that only differs in case.

This causes the files that change case from one update to the other to not be deleted at all, and both `$old` and `$new` versions stay around.

On a case-insensitive filesystem, I assume that this would avoid deleting the file from `$new` as the command to delete the file from `$old` would also remove the one from `$new`, right?

In that case, the test that you wrote could only possibly work on a case-insensitive filesystem (which is not part of our testing matrix right now), and the system will contain both `$old` and `$new` files on all case-sensitive filesystems.

I'm not sure that is the outcome we want. I think the deletion of `$old` should still always happen on case-sensitive filesystems. So we would only want to do case-insensitive comparisons when on a case-insensitive filesystem. But that is really hard to detect reliably, from what I gather...

Is my reasoning correct so far?

**Comment by pbiron** [comment]
Alain, I'm heads down in some client work at the moment.  As soon as I surface from that I'll have to refresh my memory about how the PR was intended to work and will respond in more detail.

**Comment by schlessera** [comment]
I tried to adapt the logic from Joomla to deal with this situation. However, their scenario is slightly different: they have a list of `$old => $new  mappings`. We only have a list of old filepaths and a list of new filepaths, without a mapping to know which ones represent the same thing even though they differ in name.

Now, how can we best test this to ensure it actually works on a case-insensitive filesystem? My local system as well as GHA run on Ubuntu.

**Comment by pbiron** [comment]
I'll test it in my Windows/WAMP environment this evening when I'm done working for the day

File: src/Core_Command.php [hunk 1]
```diff
@@ -1372,7 +1372,69 @@ private function cleanup_extra_files( $version_from, $version_to, $locale, $inse
...
+		// Compare the files from the old version and the new version in a case-insensitive manner,
+		// to prevent files being incorrectly deleted on systems with case-insensitive filesystems
+		// when core changes the case of filenames.
+		// The main logic for this was taken from the Joomla project and adapted for WP.
+		// See: https://github.com/joomla/joomla-cms/blob/bb5368c7ef9c20270e6e9fcc4b364cd0849082a5/administrator/components/com_admin/script.php#L8158
+
+		$old_filepaths = array_keys( $old_checksums );
+		$new_filepaths = array_keys( $new_checksums );
+
+		$new_filepaths = array_combine( array_map( 'strtolower', $new_filepaths ), $new_filepaths );
```
Comment (wojsmol) [inline src/Core_Command.php:1384]: AFAIK if we pass new checksums trough `strtolower` then we lose case information witch are very important in this code.

Comment (schlessera) [reply src/Core_Command.php:1384]: This line builds an array that use the lowercased string as the key and the original string as the value, to help with mapping.

Comment (schlessera) [reply src/Core_Command.php:1384]: Note the `array_combine()`.

So we'll go from this:
```
[
   0 => 'Abc',
   1 => 'BCD',
]
```
to this:
```
[
   'abc' => 'Abc',
   'bcd' => 'BCD',
]
```

**Review by wojsmol — APPROVED** [review]

**Pull request merged**